src/librustc_lint/builtin.rs

```diff
@@ -64,9 +64,25 @@ declare_lint! {
 
 declare_lint_pass!(WhileTrue => [WHILE_TRUE]);
 
+fn as_while_cond(expr: &hir::Expr) -> Option<&hir::Expr> {
```

Comment:
we could also make the lint an early lint pass.

Reply:
Good idea. I'll get that a try.

Reply:
Done.